### ISSUE TITLE
IndexError: index 0 is out of bounds for axis 0 with size 0

### ISSUE THREAD

**Issue by j3mdamas** [choderalab/yank#626]
Hi guys,

I am using the [yank-examples](https://github.com/choderalab/yank-examples), namely the hydration free energy of phenol (https://github.com/choderalab/yank-examples/tree/master/examples/hydration/phenol)

I installed yank using conda, and followed the tutorial, and it errored with:

```
2017-01-17 14:44:41,311: Deconvoluting replicas...
2017-01-17 14:44:41,312: Done.
2017-01-17 14:44:41,312: number of uncorrelated samples:
2017-01-17 14:44:41,312: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
2017-01-17 14:44:41,313:
2017-01-17 14:44:41,313: Computing free energy differences...
Traceback (most recent call last):
  File "/home/joao/miniconda3/bin/yank", line 11, in <module>
    load_entry_point('yank==0.14.1', 'console_scripts', 'yank')()
  File "/home/joao/miniconda3/lib/python3.5/site-packages/yank/cli.py", line 74, in main
    dispatched = getattr(commands, command).dispatch(command_args)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/yank/commands/analyze.py", line 64, in dispatch
    analyze.analyze(args['--store'])
  File "/home/joao/miniconda3/lib/python3.5/site-packages/yank/analyze.py", line 506, in analyze
    mbar = initialize_MBAR(ncfile, u_kln=u_kln, N_k=N_k)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/yank/analyze.py", line 242, in initialize_MBAR
    mbar = MBAR(u_kln, N_k)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/pymbar/mbar.py", line 264, in __init__
    self.f_k = mbar_solvers.solve_mbar_with_subsampling(self.u_kn, self.N_k, self.f_k, solver_protocol, subsampling_protocol, subsampling, x_kindices=self.x_kindices)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/pymbar/mbar_solvers.py", line 481, in solve_mbar_with_subsampling
    f_k_nonzero, all_results = solve_mbar(u_kn_subsampled, N_k_subsampled, f_k[states_with_samples], solver_protocol=subsampling_protocol)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/pymbar/mbar_solvers.py", line 371, in solve_mbar
    f_k_nonzero, results = solve_mbar_once(u_kn_nonzero, N_k_nonzero, f_k_nonzero, **options)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/pymbar/mbar_solvers.py", line 305, in solve_mbar_once
    f_k_nonzero = f_k_nonzero - f_k_nonzero[0]  # Work with reduced dimensions with f_k[0] := 0
IndexError: index 0 is out of bounds for axis 0 with size 0
```

I also attach the `experiments.log` file. Everything I tried was out of the box.

Thanks,
João

cc: @gerardbcn

[experiments.txt](https://github.com/choderalab/yank/files/714110/experiments.txt)

**Comment by Lnaden** [comment]
There are a few things that strike me as odd here. The first is that according to the error you posted here, MBAR detected 0 decorated samples across the entire simulation. Effectively its trying to run statistics on N=0, which it never expects and thus the error. 

That leads me to the the 2nd problem in that the logfile you provided appears to show the simulation throwing NaN's a couple times before giving up. We have been trying to work with OpenMM's lead dev to trace down why we see NaN's occurring and we think we may have finally gotten them all worked out in recent development versions.

What version of OpenMM are you running right now? `conda list openmm`
What command did you issue to start the analysis? The N=0 is worrying me since I think the minimum the de-correlation algorithm can return is 1.

I think we can fix part of the problem with a new version of OpenMM, its the 2nd part that may take more debugging.

**Comment by j3mdamas** [comment]
Hi Levi,

Thanks for the prompt response. I am using openmm 7.0.1 (as a matter of clarity, it's not the one in the omnia channel, but one we have on acellera channel, which is a sync of 7.0.1 https://anaconda.org/acellera/openmm).

Yeah, I saw the NaNs on the log, and I have seen that there has been some NaN issue reporting earlier, but I didn't think it was related.

I followed what is in the tutorial, which actually was just running [`run-explicit.sh`](https://github.com/choderalab/yank-examples/blob/master/examples/hydration/phenol/run-explicit.sh).

This is out-of-the-box tutorial/example. I didn't do any modification.

João

**Comment by Lnaden** [comment]
The NaN's are very much related sadly, OpenMM 7.0.1 has some issues with the neighbor lists which cause NaN's in YANK. It also looks like your sync has 7.1.0 which I think still had the outstanding NaN issues, there is a beta for 7.2.0 available for linux only through the `omnia` conda channel. What OS are you running this on? 

In the meantime, I will try to replicate the analyze bug as best I can here. Can you provide what GPU you are using (and CUDA version if its an nvidia card). The OS would be helpful for this as well.

**Comment by j3mdamas** [comment]
```
$ cat /etc/centos-release
CentOS Linux release 7.2.1511 (Core)
```

```
$ nvidia-smi
Wed Jan 18 16:28:23 2017       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 367.44                 Driver Version: 367.44                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+======================|
|   0  GeForce GTX 780     On   | 0000:02:00.0     N/A |                  N/A |
| 26%   37C    P8    N/A /  N/A |   1091MiB /  3017MiB |     N/A      Default |
+-------------------------------+----------------------+----------------------+
```

True, I was using an un-updated version. Sorry. I will try with 7.1.0 first, and then try with the 7.2.0 beta. I will report back.

Once again, thanks!

**Comment by jchodera** [comment]
For now, please try the `7.2.0` version, which you can get via:
```
conda install --yes -c omnia/label/dev openmm==7.2.0
```
since this contains the NaN bugfixes.

**Comment by j3mdamas** [comment]
Ok John. I am going to try now

```
$ conda list openmm
# packages in environment at /home/joao/miniconda3:
#
openmm                    7.2.0                    py35_0    omnia/label/dev
openmmtools               0.8.3                    py35_1    omnia
```

**Comment by Lnaden** [comment]
@j3mdamas I figured out what happened. Your solvent phase crashed with a NaN resulting in the simulations stopping, however, YANK actually initializes both NetCDF files (where data is saved) before starting either phase of the simulation. 

* Your solvent phase ran, crashed, but had partial data which probably went through MBAR just fine. 
* However, the vacuum phase had no samples, but had the initialized the file, so N=0 samples were fed in for the phase, which MBAR cannot (and should not) handle, resulting in the crash.

I think the update to the 7.2.0 beta of OpenMM should resolve the nans, allowing simulations to run, and avoiding this issue. Any luck with the new version?

**Comment by j3mdamas** [comment]
Same error:

```
2017-01-18 19:26:36,433: Reading energies...
2017-01-18 19:26:36,433: Done.
2017-01-18 19:26:36,433: Deconvoluting replicas...
2017-01-18 19:26:36,434: Done.
2017-01-18 19:26:36,434: number of uncorrelated samples:
2017-01-18 19:26:36,434: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0]
2017-01-18 19:26:36,435:
2017-01-18 19:26:36,435: Computing free energy differences...
Traceback (most recent call last):
  File "/home/joao/miniconda3/bin/yank", line 11, in <module>
    load_entry_point('yank==0.14.1', 'console_scripts', 'yank')()
  File "/home/joao/miniconda3/lib/python3.5/site-packages/yank/cli.py", line 74, in main
    dispatched = getattr(commands, command).dispatch(command_args)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/yank/commands/analyze.py", line 64, in dispatch
    analyze.analyze(args['--store'])
  File "/home/joao/miniconda3/lib/python3.5/site-packages/yank/analyze.py", line 506, in analyze
    mbar = initialize_MBAR(ncfile, u_kln=u_kln, N_k=N_k)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/yank/analyze.py", line 242, in initialize_MBAR
    mbar = MBAR(u_kln, N_k)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/pymbar/mbar.py", line 264, in __init__
    self.f_k = mbar_solvers.solve_mbar_with_subsampling(self.u_kn, self.N_k, self.f_k, solver_protocol, subsampling_protocol, subsampling, x_kindices=self.x_kindices)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/pymbar/mbar_solvers.py", line 481, in solve_mbar_with_subsampling
    f_k_nonzero, all_results = solve_mbar(u_kn_subsampled, N_k_subsampled, f_k[states_with_samples], solver_protocol=subsampling_protocol)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/pymbar/mbar_solvers.py", line 371, in solve_mbar
    f_k_nonzero, results = solve_mbar_once(u_kn_nonzero, N_k_nonzero, f_k_nonzero, **options)
  File "/home/joao/miniconda3/lib/python3.5/site-packages/pymbar/mbar_solvers.py", line 305, in solve_mbar_once
    f_k_nonzero = f_k_nonzero - f_k_nonzero[0]  # Work with reduced dimensions with f_k[0] := 0
IndexError: index 0 is out of bounds for axis 0 with size 0
```

New `experiments.log` file, still with NaN's (but no retrying?)

[experiments.txt](https://github.com/choderalab/yank/files/717119/experiments.txt)

**Comment by j3mdamas** [comment]
@Lnaden thanks for your time. It would be great if we could put the example to work.

**Comment by Lnaden** [comment]
The log file you provided does not show an actual crash, it shows a few NaN's, but it just seems to stop at the end, so I don't know what caused the termination, but I am guessing a series of NaN's. The error you posted is the same problem as before in that the 2nd phase of the simulation part did not run since the first phase crashed, so the analysis step is still looking at a blank file.

@jchodera could it be that the version labeled 7.2.0 is actually using old code still? The Anaconda site says that the files for linux 7.2 were only uploaded about 23hrs ago (from this post). My personal   current install of "7.2.0" has a `version.py` file which points to a github hash from late December, before all the neighbor list bugs went in. Since I don't know quite how the `openmm-dev` builds are uploaded, is it possible that @j3mdamas has a version that does not have the bug fixes? Is the version.py file a decent way to check that?

**Comment by j3mdamas** [comment]
Well, I don't have other outputs. Am I doing anything wrong?

I've done the installation John suggested. If there's a different version to install, please tell me, and I'll install and test it. We want to calculate the free energy of solvation of many compounds and it would be really helpful to have this protocol working.

**Comment by Lnaden** [comment]
No, you're not doing anything wrong. The `conda install` process makes installing easy, but slightly harder to debug when there are subtle version differences and file updates. We just need to make sure we get you the right version to test with. 

I am also testing locally to try and replicate the issue here to further debug. Can you provide me the text of the file from your conda environment's python `site-packages/simtk/openmm/version.py`?

**Comment by j3mdamas** [comment]
You're right!

```
$ cat ~/miniconda3/lib/python3.5/site-packages/simtk/openmm/version.py

# THIS FILE IS GENERATED FROM OPENMM SETUP.PY
short_version = '7.1.0'
version = '7.1.0'
full_version = '7.1.0.dev-9500f3a'
git_revision = '9500f3afc6f870ecbadfc5e6ab3679fe32b5ac7b'
release = False
openmm_library_path = r'/home/joao/miniconda3/lib'

if not release:
    version = full_version
```

How do I fix it?

**Comment by Lnaden** [comment]
If the `version.py` is to believed, then you have the most recent commit which should have had all the NaN bugfixes in. The fact that you are still seeing NaN's indicate that there is still a bug we have not fully diagnosed yet. We (YANK devs) will need to coordinate with the OpenMM devs to figure out the bug, unless there is something I am missing.

**Comment by j3mdamas** [comment]
So, do you guys solve this within you, or should I go to the `openmm` issues page?

Also, can everybody (you, for example) run this example? When you said:

> I am also testing locally to try and replicate the issue here to further debug

I got the impression it is working for you. If it is so, what is your settings? Do you manually check-out the openmm repo or something? We are looking for a fast solution, even if not officially released.

**Comment by jchodera** [comment]
> If the version.py is to believed, then you have the most recent commit which should have had all the NaN bugfixes in. The fact that you are still seeing NaN's indicate that there is still a bug we have not fully diagnosed yet. We (YANK devs) will need to coordinate with the OpenMM devs to figure out the bug, unless there is something I am missing.

Can you guys try to reproduce this locally? It is important to see if we can sort out those NaN issues ASAP since [OpenMM is about to go into 7.1 release candidate](https://github.com/pandegroup/openmm/issues/1718#issuecomment-273627156).

**Comment by Lnaden** [comment]
Keep the issue here, I am going to have to try and make a test case to replicate the issue and try before I can pass on the information.

> I got the impression it is working for you. If it is so, what is your settings? 

Its not really a fair comparison, I was also seeing NaN's, but I had a slightly older version of the code that did not have the supposed fixes in. 

> Do you manually check-out the openmm repo or something? 

I use `conda` to get my versions, I forgot that there is a nightly push to the `omnia/label/dev` channel and did not update. You can install OpenMM from source if you need to, but with the nightly conda build its not really required.

>  We are looking for a fast solution, even if not officially released.

I'm not sure how fast the solution will be since I first have to trap the error in a repeatable way. Sorry for the inconvenience on that part. What are you looking to do that you need a fast solution for if I may ask?

> Can you guys try to reproduce this locally? It is important to see if we can sort out those NaN issues ASAP

@jchodera  yes, that is now my highest priority since I knew that was coming down the works. I had not had issues on the example runs I was doing, but I also did not check every log file as I figured the NaN's would halt the process eventually.

**Comment by j3mdamas** [comment]
The fast solution was if there was a combination of factors that could be used to get a working, traceable version of YANK/OpenMM. We want to calculate the free energy of solvation of a library of compounds.

And no need to apologize, we undestand YANK is still in development and we appreciate the work you guys have been doing on it. Thank you.

Please, let me know if there's anything I can do to help.

João
(cc: @gerardbcn)

**Comment by j3mdamas** [comment]
@Lnaden better said: is there a known combination of version of YANK and OpenMM that works and produces correct results? If there is and you know it, can you share it? Thanks.

**Comment by jchodera** [comment]
> The fast solution was if there was a combination of factors that could be used to get a working, traceable version of YANK/OpenMM. We want to calculate the free energy of solvation of a library of compounds.

It's important to stress that we still have not officially released a version 1.0 production, validated version of YANK we recommend for use in publications. At this point, everything is still in flux, so it's entirely "use at your own risk".

**Comment by Lnaden** [comment]
At the moment, No. There is some subtle bug that causes NaN's and seemingly random intervals that we thought we finally figured out from the neighbor lists, but it appears there is something else going on we have not identified. 

As an inefficient, brute force, and possibly incorrect work around of sorts, you can try setting `resume_setup: yes` and `resume_simulation: yes` in the `explicit.yaml` file, the just rerunning `yank script --yaml=explicit.yaml` until it gets through all the iterations for both phases. I really don't recommend this and it will not work for what you are trying to do, but it in theory would get through the simulations, maybe?

**Comment by j3mdamas** [comment]
Yeah John, you are right. I did notice that YANK isn't on 1.0 yet, hence my other comment on the following line.

@Lnaden thanks, but I guess that is not a solution we are looking for.

**Comment by Lnaden** [comment]
I didn't think so. Im working on trying to trap the NaN now.

**Comment by Lnaden** [comment]
There was a fix for the OpenCL platform on OpenMM which should fix the problems you are having here. You will need to reinstall the OpenMM dev version from 

`conda install -c omnia/label/dev openmm=7.2.0`.

Give this a try and see if it solves your issues on the OpenCL platform

**Comment by j3mdamas** [comment]
I'm going to try it right now.

Report to you soon.

**Comment by jchodera** [comment]
You may need to force downgrade and then upgrade just to be sure:
```bash
# Force downgrade
conda install --yes -c omnia openmm==7.0.1
# Clean cache
conda clean -plti --yes
# Install latest
conda install --yes -c omnia/label/dev openmm==7.2.0
```

**Comment by j3mdamas** [comment]
Thanks @jchodera, I'll use your suggestion.

**Comment by j3mdamas** [comment]
After following John's instructions, I think I have the latest commit that I see on pandegroup/openmm@9567ddb304c48d336e82927adf2761e8780e9270

```
$ cat ~/miniconda3/lib/python3.5/site-packages/simtk/openmm/version.py

# THIS FILE IS GENERATED FROM OPENMM SETUP.PY
short_version = '7.1.0'
version = '7.1.0'
full_version = '7.1.0.dev-9567ddb'
git_revision = '9567ddb304c48d336e82927adf2761e8780e9270'
release = False
openmm_library_path = r'/home/joao/miniconda3/lib'

if not release:
    version = full_version
```

Going to run now.

**Comment by Lnaden** [comment]
That should be the exact version that has the OpenCL patch in it, it should be good.

**Comment by j3mdamas** [comment]
NaNs problem solved! Thanks for all your help.

For note, I got these warnings in `mbar.py`:
```
2017-01-27 21:24:00,519: Computing free energy differences...
2017-01-27 21:24:00,702: Computing covariance matrix...
/home/joao/miniconda3/lib/python3.5/site-packages/pymbar/mbar.py:1402: VisibleDeprecationWarning: using a boolean instead of an integer will result in an error in the future
  d2[(np.any(d2) < warning_cutoff)] = 0.0
/home/joao/miniconda3/lib/python3.5/site-packages/numpy/matrixlib/defmatrix.py:318: VisibleDeprecationWarning: using a boolean instead of an integer will result in an error in the future
  out = N.ndarray.__getitem__(self, index)
/home/joao/miniconda3/lib/python3.5/site-packages/pymbar/mbar.py:1403: RuntimeWarning: invalid value encountered in sqrt
  return np.sqrt(np.array(d2))
```

I have some unrelated questions, I will open another issue.

**Comment by Lnaden** [comment]
Short answer: Those warnings are fine.

Longer answer: 
The first two are some warnings that NumPy throws when you give bool indices that are not quite the same shape. For our purposes its fine.

The last warning is an artifact of how the d2 matrix is computed. That warning occurs when there is - in the d2 array, but in all but the rarest of cases, its just a floating point error on a number that is on the order of `-1E-200` or something when it should be 0. There are additional checks we have in the code for when you get a negative number in the covariance matrix that should not be 0, and those raise proper errors. We have plan for MBAR to refactor it to avoid those warnings by changing the solvers that would compute `d2`.
